documentation/examples/tail-wal/main.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"os"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/wal"
+	"gopkg.in/alecthomas/kingpin.v2"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+var (
+	a      = kingpin.New("WAL tail example", "Example program for tailing a Prometheus WAL.")
+	walDir = a.Arg("wal.dir", "Full path to a Prometheus WAL directory.").Required().String()
+	logger = log.With(log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)), "ts", log.DefaultTimestampUTC, "caller", log.DefaultCaller)
+)
+
+func decode(segment *wal.Segment) bool {
+	// todo: callum, is there an easy way to detect if r.Next() has returned the last possible record in a segment.
+	// For now just recover from the panic that is thrown if we call r.Next() when we shouldn't have
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(log.With(logger)).Log("err", r)
+			segment.Close()
+		}
+	}()
+	r := wal.NewReader(segment)
+	var (
+		dec      tsdb.RecordDecoder
+		series   []tsdb.RefSeries
+		samples  []tsdb.RefSample
+		nSeries  = 0
+		nSamples = 0
+	)
+	for {
+		ok := r.Next()
+		rec := r.Record()
```

Comment:
If you check ok before calling r.Record() I think all the defer/recover stuff can be avoided. Using a `for r.Next()` loop would also simplify everything, then just static returns (true if in the for, false after the loop)

Reply:
yep, for a while I wanted to change how `r.Next` worked to have it return an error but for now I don't think it's necessary